### ISSUE TITLE
In container, "/etc/localtime" allways points to "/usr/share/zoneinfo/UTC"

### ISSUE THREAD

**Issue by xinnj** [docker/docker#16481]
<h3>How to reproduce</h3>
[root@HBRPS002 ~]$ docker run -it -v /usr/share/zoneinfo/Asia/Shanghai:/etc/localtime:ro centos bash
[root@74bb3c95d678 /]# ls -l /etc/localtime
lrwxrwxrwx. 1 root root 23 Apr 15 22:29 /etc/localtime -> /usr/share/zoneinfo/UTC
[root@74bb3c95d678 /]# cat /etc/localtime
TZifip ~hp"^Jp$Gg%_&+(~pDTCSTTZif2
                                      ~6C)ip ~h!Iap"^J#)Cp$Gg%_&'I&(+(q~pMTCDTCST
**CST-8**

<h3>Findings</h3>
* First, the symbol link "/etc/localtime" shouldn't point to UTC zone as docker mount volumn "/usr/share/zoneinfo/Asia/Shanghai" as it.
* Second, the contents of "/usr/share/zoneinfo/UTC" is overwritten by the contents of zone "Asia/Shanghai" (CST-8). The normal contents of UTC zone file should be:

[root@HBRPS002 ~]# cat /usr/share/zoneinfo/UTC
TZif2UTCTZif2UTC
**UTC0**

The behaviour of docker v1.6 with the same command (and I think it's the correct behaviour) is:
In container, remove the symbol link "/etc/localtime", then create a new file named "/etc/localtime" and its contents are from the volumn of the host specified (for this case, it's "/usr/share/zoneinfo/Asia/Shanghai").

<h3>Causing Problem</h3>
The command "date" of Linux will read the contents of "/etc/localtime" to determine the actual timezone the system has regardless the file name pointed by "/etc/localtime". So, if you execute "date" command in the container, the correct time will be got. 
But unfortunately, java api only determine system timezone from timezone file name and doesn't care about the actual contents of the timezone file. So, if "/etc/localtime" points to "/usr/share/zoneinfo/UTC" but the contents are overwritten by "Asia/Shanghai" timezone file, java api still returns time of UTC zone though we want it returns Asia/Shanghai time.

<h3>Related Issue</h3>
I believe the root cause of this issue is same.
https://github.com/docker/docker/issues/14220

<h3>Environment:</h3>
[root@HBRPS002 ~]# uname -a
Linux HBRPS002 3.10.0-229.11.1.el7.x86_64 #1 SMP Thu Aug 6 01:06:18 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
[root@HBRPS002 ~]# docker version
Client:
 Version:      1.8.2
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   0a8c2e3
 Built:        Thu Sep 10 19:08:45 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.2
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   0a8c2e3
 Built:        Thu Sep 10 19:08:45 UTC 2015
 OS/Arch:      linux/amd64
[root@HBRPS002 ~]# docker info
Containers: 1
Images: 24
Storage Driver: devicemapper
 Pool Name: docker-202:1-401612-pool
 Pool Blocksize: 65.54 kB
 Backing Filesystem: extfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 1.817 GB
 Data Space Total: 107.4 GB
 Data Space Available: 16.62 GB
 Metadata Space Used: 2.126 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.145 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.93-RHEL7 (2015-01-28)
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.10.0-229.11.1.el7.x86_64
Operating System: CentOS Linux 7 (Core)
CPUs: 2
Total Memory: 3.452 GiB
Name: HBRPS002
ID: 7XW6:7WTF:C5EI:OK4W:KNCO:R7J5:DUGB:LWMK:XK4Z:WWSB:CTLJ:UFMF

**Comment by fudanglp** [comment]
same findings

**Comment by thaJeztah** [comment]
Yeah, I think the issue here is that mounting to the container will resolve symlinks for the mountpoint to use;

Using a minimal dockerfile to illustrate the issue;

```dockerfile
FROM alpine
RUN mkdir -p /dest && echo 'destination-file' > /dest/file.txt && ln -s /dest/file.txt /symlink.txt
```

Creating a test-directory with a file to mount into the container, and building the above Dockerfile;

```bash
mkdir repro-16481
cd repro-16481
echo 'file from host' > host.txt

echo -e "FROM alpine\nRUN mkdir -p /dest && echo 'destination-file' > /dest/file.txt && ln -s /dest/file.txt /symlink.txt" | docker build -t repro-16481 -
```

When bind-mounting the file to `/symlink.txt` inside the container, that symlink is followed, and the file is actually mounted over `/dest/file.txt`, shadowing the file that was there;

```bash
docker run -it --rm --name my-container -v ./host.txt:/symlink.txt repro-16481
/ # ls -l /symlink.txt
lrwxrwxrwx    1 root     root            14 Sep 18 11:57 /symlink.txt -> /dest/file.txt
/ # cat symlink.txt
file from host
/ # cat /dest/file.txt
file from host
```

So, the container's config has the `/symlink.txt` as destination;


```bash
docker inspect --format='{{json .Mounts}}' my-container
[{"Type":"bind","Source":"/Users/thajeztah/Projects/test/repro-16481/host.txt","Destination":"/symlink.txt","Mode":"","RW":true,"Propagation":"rprivate"}]
```

Which is also what's in the OCI spec that's sent to the OCI runtime (runc);

```bash
docker inspect --format='{{.ID}}' my-container
d270f2fe277bd7a5751053cc8b5fd7cda3db210517ba1019184daccb45304359

# get the OCI spec from containerd, and find the mount
ctr -n moby c info d270f2fe277bd7a5751053cc8b5fd7cda3db210517ba1019184daccb45304359 | jq -c '[ .Spec.mounts[] |select( .destination | contains("/symlink.txt")) ]'
[{"destination":"/symlink.txt","type":"bind","source":"/root/repro-16481/host.txt","options":["rbind","rprivate"]}]
```

So resolving the symlink is done by the OCI runtime. I'm not 100% sure if this is documented in the spec, but it's consistent with how `mount` handles this on Linux; here's reproducing the same when running inside a container (I used `--privileged` to get permissions to do so, but not recommended to run containers with `--privileged`!)

```bash
docker run -it --rm --privileged repro-16481
/ # ls -l /symlink.txt
lrwxrwxrwx    1 root     root            14 Sep 18 12:18 /symlink.txt -> /dest/file.txt
/ # echo 'my-file' > myfile.txt
/ # mount -o bind /myfile.txt /symlink.txt
/ # cat symlink.txt
my-file
/ # cat /dest/file.txt
my-file
````

**Comment by thaJeztah** [comment]
Did a quick search in the runtime-spec repository, and it looks like there's an existing ticket about ambiguity on how runtimes are expected to handle symlinks; 

- https://github.com/opencontainers/runtime-spec/issues/1165

**Comment by fudanglp** [comment]
I think there are some special treatment for `/etc/localtime`

it points to `/etc/localtime -> /usr/share/zoneinfo/Etc/UTC` even if I mapped a concrete file such as `-v ./README.md:/etc/localtime`

![image](https://github.com/user-attachments/assets/6101f246-c88e-40ac-8bea-7ea42eaaee0d)

**Comment by fudanglp** [comment]
sorry, I find `ubuntu:focal` is fine, but `ros:noetic` shows the symptom.
let me check its entrypoint

**Comment by fudanglp** [comment]
https://github.com/osrf/docker_images/blob/df19ab7d5993d3b78a908362cdcd1479a8e78b35/ros/noetic/ubuntu/focal/ros-core/Dockerfile#L7

Dockerfile did `ln -s /usr/share/zoneinfo/Etc/UTC /etc/localtime`, But I'm wonder how to overwrite it by `--volume` when `docker run`

**Comment by fudanglp** [comment]
[tzlocal](https://pypi.org/project/tzlocal/) will be confused by the symlink.

> You can also use tzlocal to get the name of your local timezone, but only if your system is configured to make that possible. tzlocal looks for the timezone name in /etc/timezone, /var/db/zoneinfo, /etc/sysconfig/clock and /etc/conf.d/clock. If your /etc/localtime is a symlink it can also extract the name from that symlink.

**Comment by tianon** [comment]
The title of this issue is confusing (and off-topic!) -- `/etc/localtime` and `/usr/share/zoneinfo/UTC` inside containers are not things provided by or controlled by Moby / the Docker Engine, but rather things provided by the image in use.

To be very explicit/clear, the only thing on-topic to discuss _here_ is whether bind mounting from the host onto the path of a symlink inside the container/image should replace the symlink with a bind mount (thus no more symlink) or if it should follow the symlink and replace the _target_ with a bind-mount (perhaps even recursively).

From what I can see, the behavior is currently (and perhaps has been for ~11 years? hard to test now thanks to cgroupsv1 vs cgroupsv2) is the latter, which I agree is surprising, but also I worry that at this point, we're very strongly in the territory of [Hyrum's Law](https://www.hyrumslaw.com/) and wonder whether this is a behavior we can actually change/fix, or whether this is something we'd need to add a new configurable flag for if we wanted to implement it (and thus it would only be available for `--mount`):

```console
$ docker run --rm --mount type=bind,src=/usr/share/zoneinfo/America/Los_Angeles,dst=/etc/localtime,ro debian:bookworm-slim mount | grep -E 'localtime|zoneinfo'
/dev/nvme0n1p3 on /usr/share/zoneinfo/Etc/UTC type ext4 (ro,noatime)
```

**Comment by thaJeztah** [comment]
Yes, we can probably update the title of this ticket (the timezone part is an effect of this specific example).

I stumbled on this ticket, and recalled that I may have looked at some similar case at some point, but didn't recall the exact details and whether the resolving happened on our side or in the OCI runtime (so spent a bit of time to narrow things down).

Indeed also was curious if the OCI spec had some mention of this (couldn't find immediately), and .

I agree it's somewhat surprising behavior; admitted, `mount` handles this similar, but it feels more natural there than as part of a container's spec (which is more "declarative").

One possible option could be to introduce a "no-resolve" option in the OCI spec that controls the behavior, without breaking backward compatibility 🤔
